### PR TITLE
Immediately compile backwards graph in AOTAutograd if dynamic shapes

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #104971

Previously, we made backwards graph compilation lazy to avoid paying
for compilation if the user didn't actually end up using the backwards
graph.  This was useful in the old days when a lot of things in Inductor
didn't work and we could bypass errors this way.

However, this has a bad implication for dynamic shapes: the backwards
graph compilation can trigger extra guards, which are too late to
install in the Dynamo context if we wait until backwards is being run.
So in this PR I move us back to compiling backwards graph immediately
if we capture any SymInts for backwards.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

File: torch/_functorch/aot_autograd.py [hunk 1]
```diff
@@ -2982,48 +2990,6 @@ def backward(ctx, *flat_args):
...
-                    assert all(a is not None for a in all_args)
-                    context = torch._C._DisableAutocast if disable_amp else nullcontext
-
-                    placeholder_list = fx_placeholder_vals(bw_module)
-
-                    # saved activations can have different stride to eager if
-                    # the compiler does layout optimization. We should restride the
-                    # tensor passed in for compiling the backward graph using the
-                    # saved tensor's stride.
-                    for i in range(len(placeholder_list)):
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/_functorch/aot_autograd.py:2995]: It's not, I am working on fixing it

File: torch/_functorch/aot_autograd.py [hunk 2]
```diff
@@ -2766,17 +2766,25 @@ def aot_dispatch_autograd(flat_fn, flat_args: List[Any], aot_config: AOTConfig,
...
                 fw_module, adjusted_flat_args
             )
 
+        # NB: It's important to compile backwards ahead of time, as this may
+        # add extra guards which we need to apply to the Dynamo cache at
+        # forwards
+        with track_graph_compiling(aot_config, "backward"):
+            placeholder_list = fx_placeholder_vals(bw_module)
+
+            compiled_bw_func = aot_config.bw_compiler(
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/_functorch/aot_autograd.py:2779]: @bdhirsh Your thing is going to need true two level cache. But IMO you should just force your users to use compiled backwards in that case, which no longer has this problem. (BTW, @jansel's compiled autograd is what convinced me to do this "simpler" fix; basically if you have any complicated situation where we don't know ahead of time what the gradients will be, you instead use compiled autograd to be able to compile given full info.)

[the remark this answers is not in the document]
Comment (bdhirsh) [reply torch/_functorch/aot_autograd.py:2779]: ok, telling users to use compiled autograd when this happens sounds fair! (still need to eventually figure out how to teach compiled autograd how to add extra bw guards)

**Comment by ezyang** [comment]
If there are no dynamic shapes, I restore the old behavior of lazy compilation to shut up some TorchScript failures. If eager backwards compilation fails, I suppress it, which gets us through some failures in our test suite where our backwards dynamic codegen doesn't actually work.

File: torch/_functorch/aot_autograd.py [hunk 3]
```diff
@@ -2795,18 +2795,81 @@ def aot_dispatch_autograd(flat_fn, flat_args: List[Any], aot_config: AOTConfig,
...
 
-
             with TracingContext.report_output_strides() as fwd_output_strides:
                 compiled_fw_func = aot_config.fw_compiler(
                     fw_module, adjusted_flat_args
                 )
 
+        # NB: It's important to compile backwards ahead of time, as this may
+        # add extra guards which we need to apply to the Dynamo cache at
+        # forwards
```
Comment (voznesenskym) [inline torch/_functorch/aot_autograd.py:2805]: yea

File: torch/_inductor/utils.py [hunk 4]
```diff
@@ -700,9 +700,10 @@ def patched_compile_to_module(self):
 
 def run_and_get_triton_code(fn, *args, **kwargs):
     _, source_codes = run_and_get_code(fn, *args, **kwargs)
+    # Can have two outputs if backwards was eagerly compiled
```
Comment (voznesenskym) [inline torch/_inductor/utils.py:703]: Can we store a flag to drive if this should be exactly 1, or of (1, 2)?

File: test/inductor/test_torchinductor.py [hunk 5]
```diff
@@ -2802,7 +2802,8 @@ def test_layer_norm(self):
             torch.nn.ReLU(),
         )
         m.eval()
-        self.common(m, (torch.randn([16, 32]),), check_lowp=False)
+        with torch.no_grad():
```
Comment (voznesenskym) [inline test/inductor/test_torchinductor.py:2805]: por que this change

File: torch/_functorch/aot_autograd.py [hunk 6]
```diff
@@ -2795,18 +2795,81 @@ def aot_dispatch_autograd(flat_fn, flat_args: List[Any], aot_config: AOTConfig,
...
+            placeholder_list = fx_placeholder_vals(bw_module)
+
+            forward_saved_for_backwards_strides = None
+            if fwd_output_strides is not None:
+                forward_saved_for_backwards_strides = fwd_output_strides[fw_metadata.tensors_saved_for_backwards_slice]
+
+            # saved activations can have different stride to eager if
+            # the compiler does layout optimization. We should restride the
+            # tensor passed in for compiling the backward graph using the
+            # saved tensor's stride.
```
Comment (voznesenskym) [inline torch/_functorch/aot_autograd.py:2816]: I would stamp but I don't know the nuances of strides well enough.

File: torch/_inductor/utils.py (hunk 4 above)
Comment (ezyang) [reply torch/_inductor/utils.py:703]: It's very awkward, because the current implementation will attempt to compile backwards, and if backwards failed to compile, suppress the error and return anyway. So there isn't really a clear delineation.

File: test/inductor/test_torchinductor.py (hunk 5 above)
Comment (ezyang) [reply test/inductor/test_torchinductor.py:2805]: Layer norm's backward compilation doesn't work, so the no grad forces us not to attempt compile it

**Comment by ezyang** [comment]
This is ready to go, just waiting for review.

File: torch/_functorch/aot_autograd.py [hunk 7]
```diff
@@ -2795,18 +2795,81 @@ def aot_dispatch_autograd(flat_fn, flat_args: List[Any], aot_config: AOTConfig,
...
+                with context():
+                    try:
+                        compiled_bw_func = aot_config.bw_compiler(
+                            bw_module, placeholder_list
+                        )
+                    except Exception:
+                        log.warning(
+                            "failed to eagerly compile backwards for dynamic, suppressing in case backwards not needed",
+                            exc_info=True
+                        )
```
Comment (eellison) [inline torch/_functorch/aot_autograd.py:2866]: What necessitates this ? It would be nice to not land with a try-catch.

Comment (ezyang) [reply torch/_functorch/aot_autograd.py:2866]: Per operator backwards dynamic codegen is still buggy af. Sample run: https://hud.pytorch.org/pytorch/pytorch/pull/104971?sha=b0bb7782a9eb8b85acbcb9ffe1835c07c6dc9f80

I'm not actually suppressing anything. If you actually try to compile backwards, we will try to compile again and THEN fail. If this suppression works out, it just means you didn't actually need the backwards graph at all.

File: torch/_functorch/aot_autograd.py [hunk 8]
```diff
@@ -2795,18 +2795,81 @@ def aot_dispatch_autograd(flat_fn, flat_args: List[Any], aot_config: AOTConfig,
...
+
+            forward_saved_for_backwards_strides = None
+            if fwd_output_strides is not None:
+                forward_saved_for_backwards_strides = fwd_output_strides[fw_metadata.tensors_saved_for_backwards_slice]
+
+            # saved activations can have different stride to eager if
+            # the compiler does layout optimization. We should restride the
+            # tensor passed in for compiling the backward graph using the
+            # saved tensor's stride.
+            for i in range(len(placeholder_list)):
```
Comment (Chillee) [inline torch/_functorch/aot_autograd.py:2817]: I'm assuming this is all code movement?

**Review by Chillee — APPROVED** [review]
iiuc, this is mostly just code movement?

File: torch/_functorch/aot_autograd.py (hunk 8 above)
Comment (ezyang) [reply torch/_functorch/aot_autograd.py:2817]: Yep
